### PR TITLE
fix: handle using compiler with React 18

### LINKED ISSUE #328: The 4.3.0 version break React compiler setup
### Describe the bug

update to the 4.3.0 and run vite dev server will catch an error:
```sh
error when starting dev server:
Error: Missing "./compiler-runtime" specifier in "react" package
    at e (file:///project/workspace/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:47597:25)
    at n (file:///project/workspace/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:47597:627)
    at o (file:///project/workspace/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:47597:1297)
    at resolveExportsOrImports (file:///project/workspace/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48287:20)
    at resolveDeepImport (file:///project/workspace/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48306:31)
    at tryNodeResolve (file:///project/workspace/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48031:20)
    at Context.resolveId (file:///project/workspace/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:47781:28)
    at Object.resolveId (file:///project/workspace/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:51063:64)
    at async file:///project/workspace/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:68310:21
    at async file:///project/workspace/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:51735:20
```

### Reproduction

https://codesandbox.io/p/devbox/react-18-react-compiler-tdrjxf?file=%2Fpackage.json%3A15%2C35

### Steps to reproduce

make sure "@vitejs/plugin-react" version is "4.3.0" in package.json, run `yarn install` and `yarn dev`.

to compare with the version "4.2.1", rewrite the "@vitejs/plugin-react" version to "4.2.1" in package.json, and re-run `yarn install` and `yarn dev`

### System Info

```shell
System:
OS: Linux 6.1 Debian GNU/Linux 12 (bookworm) 12 (bookworm)
CPU: (2) x64 AMD EPYC
Memory: 989.90 MB / 4.01 GB
Container: Yes
Shell: Unknown
Binaries:
Node: 20.9.0 - /usr/local/bin/node
Yarn: 1.22.19 - /usr/local/bin/yarn npm: 9.8.1 - /usr/local/bin/npm
pnpm: 8.10.2 - /usr/local/share/npm-global/bin/pnpm nomPackages:
@vitejs/plugin-react: 4.3.0 => 4.3.0
vite: 5.2.11 => 5.2.11
```

### Used Package Manager

yarn

### Logs

yarn run v1.22.22
$ vite
error when starting dev server:
Error: Missing "./compiler-runtime" specifier in "react" package
    at e (file:///Users/hu/Developer/react-comipler-demo-2/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:47597:25)
    at n (file:///Users/hu/Developer/react-comipler-demo-2/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:47597:627)
    at o (file:///Users/hu/Developer/react-comipler-demo-2/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:47597:1297)
    at resolveExportsOrImports (file:///Users/hu/Developer/react-comipler-demo-2/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48287:20)
    at resolveDeepImport (file:///Users/hu/Developer/react-comipler-demo-2/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48306:31)
    at tryNodeResolve (file:///Users/hu/Developer/react-comipler-demo-2/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48031:20)
    at Context.resolveId (file:///Users/hu/Developer/react-comipler-demo-2/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:47781:28)
    at Object.resolveId (file:///Users/hu/Developer/react-comipler-demo-2/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:51063:64)
    at async file:///Users/hu/Developer/react-comipler-demo-2/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:68310:21
    at async file:///Users/hu/Developer/react-comipler-demo-2/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:51735:20
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite-plugin-react/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite-plugin-react/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite-plugin-react/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite-plugin-react/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by ArnaudBarre**
Fix #328

### CODE REVIEW

**Comment by ArnaudBarre** [comment]
I'm surprised `vite build` just let the `require` from a local package without `type: module` pass though like this, feel like a bug in Vite.

File: playground/compiler-react-18/lib/react-compiler-runtime/index.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+const React = require('react')
```
Comment (hi-ogawa) [inline playground/compiler-react-18/lib/react-compiler-runtime/index.js]:
```suggestion
import React from 'react'
```

File: playground/compiler-react-18/lib/react-compiler-runtime/package.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "name": "react-compiler-runtime",
+  "private": true,
+  "main": "index.js",
```
Comment (hi-ogawa) [inline playground/compiler-react-18/lib/react-compiler-runtime/package.json:4]:
```suggestion
  "type": "module",
  "exports": "./index.js",
```

**Review by hi-ogawa — COMMENTED** [review]
> I'm surprised `vite build` just let the `require` from a local package without `type: module` pass though like this, feel like a bug in Vite.

I just ran locally and I get what you mean. I think I've seen something similar when my transform plugin accidentally injected `export ...` into react cjs package which had `require`. (I actually I wasn't sure at that time, but now seeing this issue, I'm guessing that was the case.)

Not sure why React team wrote that, but I think we can simply pick esm or cjs? I'll comment what I tested locally.

**Comment by ArnaudBarre** [comment]
Yeah for sure using ESM is good solution, but the official gist provided by @poteto uses `require` so other people maybe have trouble with it?

@poteto are you ok with changing your gist to use ESM syntax? I think all frameworks/toolchains support it

**Comment by poteto** [comment]
yeah either works, the gist is just a reference

**Comment by wong2** [comment]
Any updates?

**Comment by ArnaudBarre** [comment]
@poteto yeah I know but default matters and I think that 9 years after ESM being spec, using them for client code will be better than an import style that never worked in browsers.

**Pull request merged**